### ISSUE TITLE
Array.raw should take a reference

### ISSUE THREAD

**Issue by hellerve** [carp-lang/Carp#591]
Currently, the type signature of `Array.raw` is `(Fn [(Array a)] (Ptr a))`. I believe it should be `(Fn [(Ref (Array a))] (Ptr a))` instead. The reason I believe that would be better is that it’s currently impossible to interface with side-effecting functions that make use of the raw memory and then getting it back in Carp. While this is supremely unsafe, it is also how a bunch of low-level C libraries operate—writing into preallocated buffers rather than allocating and returning their own. I think we should provide the means to interoperate with those libraries.

What do y’all think?

Cheers

**Comment by hellerve** [comment]
Not really! Take crypto libraries, for instance: they generate a key for you, but you’ll have to generate the buffer. In C:

```
uint8_t* buf = malloc(BUF_SIZE);
my_fake_lib_generate_key(buf);
// and from here on we can use the data in buf
```

In carp, this is basically impossible.

My case in point is my wrapper for [libhydrogen](https://github.com/jedisct1/libhydrogen/). All of the functions follow this scheme, and currently I’m using a custom C data type in Carp to support what we’re trying to do (you can look at the code [here](https://github.com/carpentry-org/hydrogen.carp)).

In short, moving the array to C does not say anything about the lifetime of the value, IMO.

**Comment by jacereda** [comment]
Maybe both variants make sense and we could have `raw!` in addition to `raw`?

**Comment by hellerve** [comment]
Actually, that’s true. I was thinking that this should be possible because it’s strictly more powerful, but the added safety benefit of the current behavior is probably worth keeping it!

**Comment by eriksvedang** [comment]
Well, it's a dangerous function either way since the current behaviour requires you to manage memory manually. Having both seems potentially useful, yes. But I think the new name should explain what makes it different a bit more.

**Comment by jacereda** [comment]
I proposed `raw!` because the `!` seems to be used in several places to
imply mutation, but I'm not particularly attached to it.

**Comment by hellerve** [comment]
How about `unsafe-raw`, then?

**Comment by eriksvedang** [comment]
Sounds like a decent name.

**Comment by GrayJack** [comment]
So what was the final decision?
 - Rename `raw` to `unsafe-raw`
 - Create `unsafe-raw!` that have this type signature `(Fn [(Ref (Array a))] (Ptr a))`

That's correct?

**Comment by hellerve** [comment]
Not quite; `raw` will stay, and `unsafe-raw` will be there as an additional function! :)

**Comment by GrayJack** [comment]
I updated, now it's right?

**Comment by hellerve** [comment]
That sounds right to me!
